lib/routes/token.js

```diff
@@ -172,18 +256,28 @@ module.exports = {
   },
   handler: function tokenEndpoint(req, reply) {
     var params = req.payload;
-    confirmClient(params.client_id, params.client_secret).then(function() {
+    P.try(function() {
       if (params.grant_type === GRANT_AUTHORIZATION_CODE) {
-        return confirmCode(params.client_id, params.code).then(function(vals) {
-          vals.ttl = params.ttl;
-          return vals;
+        return confirmClient(params.client_id, params.client_secret)
+        .then(function() {
+          return confirmCode(params.client_id, params.code);
         });
       } else if (params.grant_type === GRANT_REFRESH_TOKEN) {
-        return confirmRefreshToken(params).then(function(vals) {
-          vals.ttl = params.ttl;
-          return vals;
+        return confirmClient(params.client_id, params.client_secret)
+        .then(function() {
+          return confirmRefreshToken(params);
         });
-      } // else our Joi validation failed us?
+      } else if (params.grant_type === GRANT_JWT) {
+        return confirmJwt(params);
+      } else {
+        // else our Joi validation failed us?
+        logger.critical('joi.grant_type', params.grant_type);
+        throw Error('unreachable');
+      }
+    })
+    .then(function(vals) {
+      vals.ttl = params.ttl;
+      return vals;
```

Comment:
This file's getting pretty long and growing a lot of concerns; I think it would benefit from a few high-level comments to assist new readers in navigating the file.